recursion/reading-recursive-code.md

```diff
@@ -0,0 +1,288 @@
+Calling pow! num: 3 exponent: 3
+Calling pow! num: 3 exponent: 2
+Calling pow! num: 3 exponent: 1
+Calling pow! num: 3 exponent: 0
+3 raised to the power of 5 is 243
+-------------
+Calling pow! num: 3 exponent: 6
+Calling pow! num: 3 exponent: 5
+Calling pow! num: 3 exponent: 4
+Calling pow! num: 3 exponent: 3
+Calling pow! num: 3 exponent: 2
+Calling pow! num: 3 exponent: 1
+Calling pow! num: 3 exponent: 0
+3 raised to the power of 6 is 729
+```
+
+We can confirm that the `pow` function was called multiple times. Each time, the problem was made smaller, and `exponent` decremented by one. Eventually, when `exponent` was `0`, it met the conditions for the base case, and returned `1`.
+
+## Infinite Recursion
+
+A function that calls itself without working towards a base case will repeat infinitely.
+
+```python
+def infinite_recursion(n):
+    return n + infinite_recursion(n)
+```
+
+A function without a base case will never have a stopping point! Consider this code which has no base case:
+
+```python
+def infinite_recursion(n):
+    return n + infinite_recursion(n - 1)
+```
+
+Executing both of these functions results in a `RecursionError`!
+
+## Stacks and Function Calls
+
+A running program has a dedicated section of memory called **the system stack**.
+
+The system stack is where method calls and local variables are stored. It's responsible for keeping track of function calls.
+
+The "stack" in "system stack" implies that it uses the _stack_ data structure. The system stack keeps track of function calls using a stack data structure.
+
+### Intro to the Stack Data Structure
+
+How does a stack organize data? A stack organizes data like a stack of plates! When we add a plate to the stack of plates, we add it to the top. When we need a plate from the stack of plates, we remove it from the top.
+
+When a stack adds a new piece of data, it goes on top. When a stack removes one piece of data, it pops off the top item.
+
+### !callout-info
+
+## Stacks Are LIFO
+
+Stacks operate in a Last-In-First-Out (LIFO) manner. They remove things in the reverse order they were added.
+
+### !end-callout
+
+## Tracing Recursive Functions and Illustrating the Stack
+
+Let's read more recursive functions and apply our new knowledge about the _system stack_ in order to visualize it. Consider the following function.
+
+```python
+def mystery(num):
+    if num <= 1:
+        return 1
+
+    return num * mystery(num-1)
+```
+
+What happens when we invoke the function when `num` is `5`?
+
+```python
+mystery(5)
+```
+
+We can begin to illustrate the _system stack_. Because the system stack keeps track of function calls, we add in the function call `mystery(5)` onto the stack.
+
+![Diagram showing the system stack as a box. There is a box labeled "mystery(5)" inside the system stack.](<../assets/recursion_reading-recursion_mystery(5).png>)  
+_Fig. Diagram showing the system stack as a box. There is a box labeled "mystery(5)" inside the system stack._
```

Comment:
I DIDN'T KNOW WE COULD WRAP THE PATH WITH <> TO INCLUDE PARENS!!!
in the alt text, it's recommended not to say this is an image/diagram. It's the alt of an img. Of course it's an image/diagram.
For the caption, only repeat the alt for ultra simple alts/captions. Otherwise, the alt text is way to specific for the caption (if the image is visible, that level of detail is unnecessary). It's more important to call out the overall take away.
Of note, readers using screen readers have access to both the alt and the caption (the caption is just text), so they should work in tandem.

The remaining alt text and captions will have similar suggestions.

```suggestion
![A box representing the call stack. There is a box labeled "mystery(5)" inside the call stack.](<../assets/recursion_reading-recursion_mystery(5).png>)  
_Fig. The top of the call stack after calling **`mystery(5)`**_
```

Reply:
Thanks for teaching me about the alt text stuff!!

I think my formatter included those lil `<>`-- I got lucky! But yes, good information :D